text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,237 @@
+    Message(MessageAddr),
+}
+```
+
+`MessageAddr` is a composite key derived from the caller's entity address and a hash of the topic name:
+```rust
+pub struct MessageAddr {
+    /// The entity addr.
+    entity_addr: AddressableEntityHash,
+    /// A 32 byte BLAKE2b hash of the name of the message topic.
+    topic_hash: TopicNameHash,
+    /// The message index. If the message index is `None`, the `MessageAddr` points to the topic control record.
+    message_index: Option<u32>,
+}
+```
+
+There are 2 types of stored values that are written to global state under this key.
+
+The control record for the message topic:
+```rust
+pub struct MessageTopicSummary {
+    /// Number of messages in this topic.
+    pub(crate) message_count: u32,
+    /// Block timestamp in which these messages were emitted.
+    pub(crate) blocktime: BlockTime,
+}
+```
+
+
+Checksum of the message payload that was emitted:
+```rust
+pub struct MessageChecksum(
+    pub  [u8; 32], //a 32 byte BLAKE2b hash of the serialized `MessagePayload` that was emitted.
+);
+```
+
+When a new message is emitted on a specified topic the `message_count` in the `MessageTopicSummary` record is incremented.
+Messages have ephemeral nature by design. In order to not increase the size of global state in an uncontrolled manner, the message count is reset to zero with every newly emitted message that belongs to a new block and the block time is updated as well. The old message checksum entries are pruned from global state.
+Off-chain applications can still query the old message checksums if required by using the state root hash of the block where the messages were emitted.
+
+### Node changes
+
+The Execution engine provides a result to the node in the form of an internal `ExecutionResult` data structure. In order to pass the message along to the node we need to extend the `ExecutionResults` to include the messages that were emitted during execution on both success and failure of the deploy.
+The node will take the messages coming off the execution results in the `contract_runtime` component and will send them through the SSE hook as part of the `TransactionProcessed` SSE event.
+The messages are extracted and only sent out the SSE event. The external facing `ExecutionResults` type from `casper-types` will not contain the messages. This is because execution results get stored in node storage. The messages have an ephemeral nature and should not be stored.
+
+The `TransactionProcessed` SSE event will contain a `messages` field that will have an array of `Message` objects:
+
+```rust
+pub struct Message {
+    /// The identity of the entity that produced the message.
+    entity_addr: AddressableEntityHash,
+    /// The payload of the message.
+    message: MessagePayload,
+    /// The name of the topic on which the message was emitted.
+    topic_name: String,
+    /// The hash of the name of the topic.
+    topic_name_hash: TopicNameHash,
+    /// Message index in the topic.
```

Comment:
`Message index in the topic` is not very useful. What is useful is having the global index for all events emitted by the contract, so I can reason about a causality.

Just to be clear I'd crate separate topics for every event type.

Reply:
We would still need the message index in the topic to be able to query the checksum. But you make a good point about cross-topic correlation of messages.
I'm thinking maybe of a use-case where you want to check that a particular "burn" message for a token id was emitted only after a "mint" message for the same id?
Having a global counter per contract would solve the sequence correlation out of the box, but at the same time it has the potential to bring some noise if a large number of unrelated topics are used.

With the current proposal you could have a separate 3rd topic used to log the sequence of the messages in the other 2 topics which should also solve the issue and have a minimal amount of noise.

Reply:
I'd introduce new index: `all_topics_index` so both are possible. `u32` is large enough for counting all events.

If the solution is having 3rd topic, I'd probably end up having just one contract-wide topic and push all the events into it. And then the whole topic feature doesn't really make sense.

Another uncovered thing is the correlation of events between subcalls. I'd like to know what triggered what (like in etherscan presents).

Consider this scenario:
```
1. Contract A emits EventA(data1).
2. Contract A calls Contract B.
3. Contract B emits EventB.
4. Contract B calls Contract (A -> B -> A)
5. Contract A emits EventA(data2).
```

I'd like to be able to easily know the sequence:
```
1. ContractA, EventA(data1)
2. ContractB, EventB
3. ContractA, EventA(data2)
```

Reply:
I added a section to the CEP related to tracking messages across contracts and topics. Please take a look here: https://github.com/casper-network/ceps/pull/88/commits/afaf653819ca6db9bf68415586ffd93aec2dd8a8

The proposal I wrote involves writing a log for the message keys that is indexed by the deploy hash. This would allow consumers to enable this functionality and track messages across contract invocations.
Please take a look.